receiver_test.go

```diff
@@ -154,40 +433,748 @@ func TestReceive_ModeSecond(t *testing.T) {
 	}
 	// subsequent dispositions should have no effect
-	// TODO: https://github.com/Azure/go-amqp/issues/76
-	/*ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
+	ctx, cancel = context.WithTimeout(context.Background(), 100*time.Millisecond)
 	err = r.AcceptMessage(ctx, msg)
 	cancel()
-	assert.NoError(t, err)*/
+	assert.NoError(t, err)
 	assert.NoError(t, client.Close())
 }
 
-func Test_ReceiveNonBlocking(t *testing.T) {
-	messagesCh := make(chan Message, 1)
-
-	receiver := &Receiver{
-		link: &link{
-			Messages:      messagesCh,
-			ReceiverReady: make(chan struct{}),
-		},
+func TestReceiveSuccessModeSecondReject(t *testing.T) {
+	const linkHandle = 0
+	deliveryID := uint32(1)
+	responder := func(req frames.FrameBody) ([]byte, error) {
+		b, err := receiverFrameHandler(ModeSecond)(req)
+		if b != nil || err != nil {
+			return b, err
+		}
+		switch ff := req.(type) {
+		case *frames.PerformFlow:
+			if *ff.NextIncomingID == deliveryID {
+				// this is the first flow frame, send our payload
+				return mocks.PerformTransfer(0, linkHandle, deliveryID, []byte("hello"))
+			}
+			// ignore future flow frames as we have no response
+			return nil, nil
+		case *frames.PerformDisposition:
+			if _, ok := ff.State.(*encoding.StateRejected); !ok {
+				return nil, fmt.Errorf("unexpected State %T", ff.State)
+			}
+			return mocks.PerformDisposition(encoding.RoleSender, 0, deliveryID, nil, &encoding.StateRejected{})
+		default:
+			return nil, fmt.Errorf("unhandled frame %T", req)
+		}
 	}
-
-	// if there are no cached messages we just return immediately - no error, no message.
-	msg, err := receiver.Prefetched(context.Background())
-	require.Nil(t, msg)
-	require.Nil(t, err)
-
-	messagesCh <- Message{
-		ApplicationProperties: map[string]interface{}{
-			"prop": "hello",
-		},
-		settled: true,
+	conn := mocks.NewNetConn(responder)
+	client, err := New(conn)
+	assert.NoError(t, err)
+	session, err := client.NewSession()
+	assert.NoError(t, err)
+	r, err := session.NewReceiver(LinkReceiverSettle(ModeSecond))
+	assert.NoError(t, err)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	msg, err := r.Receive(ctx)
+	cancel()
+	assert.NoError(t, err)
+	if c := r.link.countUnsettled(); c != 1 {
+		t.Fatalf("unexpected unsettled count %d", c)
 	}
+	// wait for the link to pause as we've consumed all available credit
+	assert.NoError(t, waitForLink(r.link, true))
+	// link credit must be zero since we only started with 1
+	if c := r.link.linkCredit; c != 0 {
+		t.Fatalf("unexpected link credit %d", c)
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), time.Second)
+	err = r.RejectMessage(ctx, msg, nil)
+	cancel()
+	assert.NoError(t, err)
+	if c := r.link.countUnsettled(); c != 0 {
+		t.Fatalf("unexpected unsettled count %d", c)
+	}
+	// perform a dummy receive with short timeout to trigger flow
+	ctx, cancel = context.WithTimeout(context.Background(), 100*time.Millisecond)
+	_, _ = r.Receive(ctx)
```

Comment:
Does the dummy receive just return (nil, nil) or similar? Wondering why we're ignoring it.

Reply:
It returns a `context.DeadlineExceedex` error which we expect as there are no messages to be received.  I admit it's a bit clunky as it really exposes implementation details (it "moves things along" as under-the-covers it reads from `l.ReceiverReady`) but I don't know of a better way to do this without changing how this works.

Reply:
I think that's fine, but it'd be good to validate that that's what happened. Otherwise we could be masking some other silly bug.

Reply:
Fixed